### PR TITLE
SW-4410 Add table for planting seasons

### PR DESCRIPTION
**Pull request by sgrimm**
Add a new database table `tracking.planting_seasons` to hold the list of planting
seasons for each planting site.

### CODE REVIEW

**Comment by sgrimm** [comment]
Current dependencies on/for this PR:
* main
  * **PR #1492** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/1492?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
    * **PR #1493** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/1493?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #1494** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/1494?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #1495** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/1495?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
          * **PR #1496** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/1496?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
            * **PR #1497** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/1497?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
              * **PR #1498** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/1498?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                * **PR #1499** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/1499?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                  * **PR #1500** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/1500?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
                    * **PR #1501** <a href="https://app.graphite.dev/github/pr/terraware/terraware-server/1501?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/terraware/terraware-server/1492?utm_source=stack-comment).

**Review by karthikbtf — APPROVED** [review]

File: src/test/kotlin/com/terraformation/backend/db/DatabaseTest.kt [hunk 1]
```diff
@@ -341,6 +346,7 @@ abstract class DatabaseTest {
   protected val organizationsDao: OrganizationsDao by lazyDao()
   protected val organizationUsersDao: OrganizationUsersDao by lazyDao()
   protected val plantingsDao: PlantingsDao by lazyDao()
+  protected val plantingSeasonsDao: PlantingSeasonsDao by lazyDao()
```
Comment (karthikbtf) [inline src/test/kotlin/com/terraformation/backend/db/DatabaseTest.kt:349]: Do we need a test to check that only one active planting season can be present?

File: src/main/resources/db/migration/0200/V227__PlantingSeasons.sql [hunk 2]
```diff
@@ -0,0 +1,19 @@
+CREATE TABLE tracking.planting_seasons (
+    id BIGINT PRIMARY KEY GENERATED BY DEFAULT AS IDENTITY,
+    planting_site_id BIGINT NOT NULL REFERENCES tracking.planting_sites ON DELETE CASCADE,
+    start_date DATE NOT NULL,
+    start_time TIMESTAMP WITH TIME ZONE NOT NULL,
+    end_date DATE NOT NULL,
+    end_time TIMESTAMP WITH TIME ZONE NOT NULL,
+    is_active BOOLEAN NOT NULL,
+
+    CHECK (start_date < end_date),
```
Comment (nickgraz) [inline src/main/resources/db/migration/0200/V227__PlantingSeasons.sql:10]: ~Should we implement the planting season restrictions here? Or are those validations elsewhere? For example "planting season must be at least 4 weeks" and "planting season is no longer than 1 year".~ Never mind, found them!

**Review by nickgraz — APPROVED** [review]

File: src/test/kotlin/com/terraformation/backend/db/DatabaseTest.kt (hunk 1 above)
Comment (sgrimm) [reply src/test/kotlin/com/terraformation/backend/db/DatabaseTest.kt:349]: I have mixed feelings about this. On the one hand, it's a behavior of the system, even if it's only a second layer of defense against something the app should never try to do. So of course it should be verified by a test that will catch someone messing up the constraint by accident in the future.

On the other hand, the exact same thing is true of the foreign-key constraint on the planting site ID column. Verifying that the database doesn't let us refer to a nonexistent planting site seems (to me) not worthwhile, especially if we apply that thinking consistently and add tests to cover _all_ the integrity constraints in the database.

Maybe the argument in favor of testing this but not the foreign key is that an exclude constraint is more esoteric than a foreign key, and the test serves as documentation for people who aren't as familiar with the syntax? Or maybe the deciding factor is that the constraint doesn't apply to all rows?

Comment (sgrimm) [reply src/test/kotlin/com/terraformation/backend/db/DatabaseTest.kt:349]: I've added a test for this, even if only so we have something more concrete to discuss.

Comment (karthikbtf) [reply src/test/kotlin/com/terraformation/backend/db/DatabaseTest.kt:349]: Thanks for sharing more context. Wasn't too sure on expectation.
Makes sense to not require tests for these cases going forward.
Would you prefer to revert the added test or keep it for discussion purposes?

**Review by karthikbtf — APPROVED** [review]

File: src/test/kotlin/com/terraformation/backend/db/DatabaseTest.kt (hunk 1 above)
Comment (sgrimm) [reply src/test/kotlin/com/terraformation/backend/db/DatabaseTest.kt:349]: I'll revert it. Thanks for the discussion, though -- definitely worth raising as a question.

Comment (karthikbtf) [reply src/test/kotlin/com/terraformation/backend/db/DatabaseTest.kt:349]: Thanks, Steve. Sounds good 👍

**Comment by sgrimm** [comment]
### Merge activity

* **Nov 27, 5:00 PM**: @sgrimm started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/terraware/terraware-server/1492).
* **Nov 27, 5:00 PM**: @sgrimm merged this pull request with [Graphite](https://app.graphite.dev/github/pr/terraware/terraware-server/1492).

**Pull request merged**